File: npy-backend/utils/mongo.py
import os
from pymongo import MongoClient
from pymongo.collection import Collection
from pymongo.database import Database
from typing import Optional
from utils.logger import logger
from utils.config import config

class MongoManager:
    _instance: Optional['MongoManager'] = None
    _client: Optional[MongoClient] = None
    _db: Optional[Database] = None

    def __init__(self):
        if MongoManager._instance is not None:
            raise Exception("This class is a singleton!")
        
        self.mongo_uri = config.MONGO_URI
        self.db_name = config.MONGO_DB_NAME
        
    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            cls._instance = MongoManager()
        return cls._instance
    
    def connect(self):
        if self._client is None:
            try:
                logger.info(f"Connecting to MongoDB at {self.mongo_uri}...")
                self._client = MongoClient(
                    self.mongo_uri,
                    serverSelectionTimeoutMS=5000,
                    connectTimeoutMS=10000,
                    socketTimeoutMS=45000,
                    maxPoolSize=config.MONGO_MAX_POOL_SIZE,
                    minPoolSize=config.MONGO_MIN_POOL_SIZE,
                    maxIdleTimeMS=config.MONGO_MAX_IDLE_TIME_MS,
                    waitQueueTimeoutMS=config.MONGO_WAIT_QUEUE_TIMEOUT_MS,
                )
                self._db = self._client[self.db_name]
                
                # Check connection
                self._client.admin.command('ping')
                logger.info("Successfully connected to MongoDB")
                
                # Ensure indexes on startup
                self._ensure_indexes()
            except Exception as e:
                error_type = type(e).__name__
                logger.error(f"Failed to connect to MongoDB ({error_type})")
                raise e
    
    def _ensure_indexes(self):
        """Create mandatory indexes at startup"""
        if self._db is not None:
            pages = self._db.pages
            pages.create_index("jobId")
            pages.create_index("url")
            pages.create_index([("jobId", 1), ("url", 1)])
            pages.create_index([("jobId", 1), ("createdAt", 1)])

            self._db.links.create_index("jobId")
            self._db.links.create_index([("jobId", 1), ("createdAt", 1)])
            self._db.sitemaps.create_index("jobId")
            self._db.sitemaps.create_index([("jobId", 1), ("createdAt", 1)])
            self._db.fields.create_index("jobId")
            self._db.fields.create_index([("jobId", 1), ("url", 1)])
            self._db.fields.create_index([("jobId", 1), ("createdAt", 1)])
            self._db.sessions.create_index([("projectId", 1), ("createdAt", -1)])
            self._db.sessions.create_index([("projectId", 1), ("status", 1), ("createdAt", -1)])
            self._db.jobs.create_index([("sessionId", 1), ("createdAt", -1)])
            self._db.projects.create_index([("userId", 1), ("status", 1), ("createdAt", -1)])
            # job_summaries: unique+sparse index on jobId.
            # An older non-unique index with the same auto-generated name
            # ("jobId_1") may exist from a previous deployment.  MongoDB
            # rejects create_index when the name matches but the spec differs
            # (code 86 IndexKeySpecsConflict), so we drop the stale index
            # first and then recreate with the correct options.
            try:
                self._db.job_summaries.create_index(
                    "jobId",
                    unique=True,
                    sparse=True,
                )
            except Exception as idx_err:
                from pymongo.errors import OperationFailure
                if isinstance(idx_err, OperationFailure) and idx_err.code == 86:
                    logger.warning(
                        "job_summaries.jobId_1 index spec conflict — dropping stale "
                        "index and recreating with unique+sparse options."
                    )
                    try:
                        self._db.job_summaries.drop_index("jobId_1")
                    except Exception:
                        pass
                    self._db.job_summaries.create_index(
                        "jobId",
                        unique=True,
                        sparse=True,
                    )
                else:
                    raise
            self._db.module_e.create_index("jobId", unique=True)
            self._db.module_f.create_index("jobId", unique=True)
            self._db.module_f.create_index([("sessionId", 1), ("updatedAt", -1)])
            self._db.module_c.create_index("jobId")
            self._db.module_c.create_index([("jobId", 1), ("url", 1)])
            self._db.module_c.create_index([("jobId", 1), ("timestamp", -1)])
            self._db.performance_audits.create_index([("jobId", 1), ("device", 1), ("runAt", -1)])
            self._db.serp_results.create_index("jobId", unique=True)
            self._db.serp_results.create_index("sessionId")

            # HTML dedup index: one entry per (project, url)
            self._db.html_dedup.create_index(
                [("project_id", 1), ("url_hash", 1)],
                unique=True,
            )
            
            logger.info("MongoDB indexes verified")

    @property
    def db(self) -> Database:
        if self._db is None:
            self.connect()
        return self._db
    
    @property
    def pages(self) -> Collection:
        return self.db.pages

    @property
    def links(self) -> Collection:
        return self.db.links

    @property
    def sitemaps(self) -> Collection:
        return self.db.sitemaps

    @property
    def job_summaries(self) -> Collection:
        return self.db.job_summaries

    @property
    def fields(self) -> Collection:
        return self.db.fields

    @property
    def schemas(self) -> Collection:
        return self.db.schemas

    @property
    def content_metrics(self) -> Collection:
        return self.db.content_metrics

    @property
    def schemas(self) -> Collection:
        return self.db.schemas

    @property
    def content_metrics(self) -> Collection:
        return self.db.content_metrics

    @property
    def module_e(self) -> Collection:
        return self.db.module_e

    @property
    def module_f(self) -> Collection:
        return self.db.module_f

    @property
    def module_c(self) -> Collection:
        return self.db.module_c

    @property
    def serp_results(self) -> Collection:
        return self.db.serp_results

    @property
    def html_dedup(self) -> Collection:
        return self.db.html_dedup

    def close(self):
        if self._client:
            self._client.close()
            self._client = None
            self._db = None
            logger.info("MongoDB connection closed")

# Global instance
mongo_manager = MongoManager.get_instance()


def get_db():
    """Return the MongoDB database instance.

    Convenience helper so callers can write::

        from utils.mongo import get_db
        db = get_db()
    """
    return mongo_manager.db

File: nnode-backend/src/config/mongo.ts
import { MongoClient, Db } from 'mongodb';
import { env } from './env';
import { logger } from '../shared/logger/logger';

let client: MongoClient | null = null;
let db: Db | null = null;
let indexesEnsured: Promise<void> | null = null;

/** Seed entries for llm_source_map if the collection is empty. */
const LLM_SOURCE_MAP_SEED = [
  { source_domain: 'chatgpt.com',          display_name: 'ChatGPT',   is_active: true },
  { source_domain: 'chat.openai.com',      display_name: 'ChatGPT',   is_active: true },
  { source_domain: 'gemini.google.com',    display_name: 'Gemini',    is_active: true },
  { source_domain: 'bard.google.com',      display_name: 'Gemini',    is_active: true },
  { source_domain: 'perplexity.ai',        display_name: 'Perplexity',is_active: true },
  { source_domain: 'claude.ai',            display_name: 'Claude',    is_active: true },
  { source_domain: 'bing.com',             display_name: 'Bing AI',   is_active: true },
  { source_domain: 'copilot.microsoft.com',display_name: 'Copilot',   is_active: true },
  { source_domain: 'you.com',              display_name: 'You.com AI',is_active: true },
  { source_domain: 'phind.com',            display_name: 'Phind',     is_active: true },
  { source_domain: 'kagi.com',             display_name: 'Kagi AI',   is_active: true },
  { source_domain: 'poe.com',              display_name: 'Poe',       is_active: true },
];

const seedLLMSourceMap = async (database: Db): Promise<void> => {
  const col = database.collection('llm_source_map');
  const count = await col.countDocuments();
  if (count === 0) {
    const now = new Date();
    await col.insertMany(LLM_SOURCE_MAP_SEED.map((e) => ({ ...e, added_at: now })));
    logger.info('llm_source_map seeded with default entries');
  }
};

const ensureMongoIndexes = async (database: Db): Promise<void> => {
  if (!indexesEnsured) {
    indexesEnsured = Promise.all([
      database.collection('llm_source_map').createIndex({ source_domain: 1 }, { unique: true }),
      database.collection('users').createIndex(
        { emailNormalized: 1 },
        {
          unique: true,
          partialFilterExpression: {
            emailNormalized: { $exists: true, $type: 'string' },
          },
        }
      ),
      database.collection('users').createIndex(
        { googleId: 1 },
        {
          unique: true,
          sparse: true,
        }
      ),
      database.collection('sessions').createIndex({ projectId: 1, createdAt: -1 }),
      database.collection('sessions').createIndex({ projectId: 1, status: 1, createdAt: -1 }),
      database.collection('jobs').createIndex({ sessionId: 1, createdAt: -1 }),
      database.collection('projects').createIndex({ userId: 1, status: 1, createdAt: -1 }),
      database.collection('pages').createIndex({ jobId: 1, createdAt: 1 }),
      database.collection('links').createIndex({ jobId: 1, createdAt: 1 }),
      database.collection('fields').createIndex({ jobId: 1, createdAt: 1 }),
      database.collection('sitemaps').createIndex({ jobId: 1, createdAt: 1 }),
      database.collection('module_c').createIndex({ jobId: 1, timestamp: -1 }),
      database.collection('performance_audits').createIndex({ jobId: 1, device: 1, runAt: -1 }),
      database.collection('module_f').createIndex({ sessionId: 1, updatedAt: -1 }),
    ])
      .then(() => {
        logger.info('MongoDB indexes verified');
      })
      .catch((error) => {
        indexesEnsured = null;
        throw error;
      });
  }

  await indexesEnsured;
};

export const connectToMongo = async (): Promise<Db> => {
  if (db) return db;

  try {
    client = new MongoClient(env.MONGO_URI, {
      serverSelectionTimeoutMS: 5000,
      connectTimeoutMS: 10000,
      socketTimeoutMS: 45000,
      maxPoolSize: env.MONGO_MAX_POOL_SIZE,
      minPoolSize: env.MONGO_MIN_POOL_SIZE,
      maxIdleTimeMS: env.MONGO_MAX_IDLE_TIME_MS,
      waitQueueTimeoutMS: env.MONGO_WAIT_QUEUE_TIMEOUT_MS,
    });
    await client.connect();

    db = client.db(env.MONGO_DB_NAME);
    await ensureMongoIndexes(db);
    await seedLLMSourceMap(db);
    logger.info('✅ Connected to MongoDB');
    return db;
  } catch (error: any) {
    logger.error(`❌ MongoDB connection error: ${error?.message || String(error)}`);
    throw error;
  }
};

export const getMongoDb = (): Db => {
  if (!db) {
    throw new Error('MongoDB not connected. Call connectToMongo first.');
  }
  return db;
};

export const closeMongoConnection = async (): Promise<void> => {
  if (client) {
    await client.close();
    client = null;
    db = null;
    indexesEnsured = null;
    logger.info('🔌 MongoDB connection closed');
  }
};
